### DISCUSSION TITLE
How to install and setup zwavejs2mqtt on a Raspberry Pi?

### DISCUSSION THREAD

**Discussion by ZeroGravityByUs** [zwave-js/zwave-js-ui#1216] [Q&A] [answered]
## Hello!
  
I'm very new to all of this, so please be patient with me!

I'm trying to install **zwavejs2mqtt** on my Raspberry Pi but I have some questions.  
 
What I have done so far:  

## Snapd
```  
sudo apt update
sudo apt install snapd
sudo reboot  
sudo snap install core
```  

## zwavejs2mqtt
 
Installing  zwavejs2mqtt
https://zwave-js.github.io/zwavejs2mqtt/#/getting-started/quick-start

```  
sudo snap install zwavejs2mqtt
sudo snap connect zwavejs2mqtt:raw-usb
sudo snap connect zwavejs2mqtt:hardware-observe
```  

## Result  
If i write `zwavejs2mqtt.help` I get the following response:  
  

```  
ERROR: ld.so: object '/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object '/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object '/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
ERROR: ld.so: object '/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
Missing plug: «serial-port»
OPTIONAL plug
Connect with: $ sudo snap connect zwavejs2mqtt:serial-port
zwavejs2mqtt ($ ($ 1 --version))

IMPORTANT! The DAEMON/SERVICE is disabled by default after installation.
You have to manually 'daemonize' it with executing the command

$ zwavejs2mqtt.enable

Set your configuration as descrived below  
  
Base configuration values:
- network.key:               361229AE4E54B9AAB10259DCCAD78572
Server configuration values: {
	"server": {
		"host": "localhost",
		"port": 8091,
		"ssl": false
	}
}

- server.ssl: Use secure communication
- server.host: IP address to bind to, e.g 127.0.0.1
- server.port: Port to reach the web interface

Session configuration values: {
	"session": {
		"cookie-secure": false,
		"secret": "2b3931cc-b41a-11eb-bce7-07294b2310ea"
	}
}

Set options with:         $ snap set zwavejs2mqtt param=key
For example:              $ snap set zwavejs2mqtt server.host=0.0.0.0

Other settings can be set in the UI after start.

Follow the log with
$ snap logs zwavejs2mqtt -f

Other commands
Disable the daemon:       $ zwavejs2mqtt.disable
Restart the daemon:       $ zwavejs2mqtt.restart
```    

## Questions

1. How do I start **zwavejs2mqtt**?   
When I write `zwavejs2mqtt.enable`  I get the following reply:  
  
```  
ERROR: ld.so: object '/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object '/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object '/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so' from /etc/ld.so.preload cannot be preloaded (cannot open shared object file): ignored.
bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
Running as root is required.
Re-run with sudo.
```  
  
If I re-run with sudo: `sudo zwavejs2mqtt.enable` I get:    
```
sudo: zwavejs2mqtt.enable: command not found
```  
  
2.  How do I configure the network? Where do I enter the information about the port etc? Should I save it in a file? If I should, what should I name that file, and where should I save it?  
  
3.  Do I need to install  **mosquitto** for  **zwavejs2mqtt** to work?  

4.  I have an Aeotec Z-stick gen5 USB-adapter, how do I "register" it with **zwavejs2mqtt**?
   

## Background
Maybe you are wondering what I'm trying to archive.
  
Well, to put it simple, I'm trying to write a program that displays a webpage on a Raspbery Pi, when a Fibaro button is pushed!  
  
I wrote a question on the Raspberry Pi forum that explains in more detail what I'm trying to do:  
https://www.raspberrypi.org/forums/viewtopic.php?f=102&t=310834&p=1858790

Then I found this article:   
https://hackernoon.com/how-to-transform-a-raspberrypi-into-a-universal-zigbee-and-z-wave-bridge-xy1ay3ymz,   
I was then recommended to use the **zwavejs2mqtt** instead (since the other zwave plugin is not being updated anymore), so now I'm trying to install **zwavejs2mqtt**, and then I will try to get it to work with Platypush if needed.

Sorry for writing so long....  
  
Any help is highly appreciated.

Thanks!

**Comment by robertsLando** [comment]
I have no experience with snap and I'm not the maintainer of snap package so I cannot help with that. Maybe @jmgiaever could help you with that

As an alternative you can consider to use the docker way

**Reply by ZeroGravityByUs** [reply]
The docker way worked! Thanks!

**Comment by jmgiaever** [comment]
Hi,

which architecture is this? 32-bit or 64-bit? And which operating system are you using?

**Comment by jmgiaever** [comment]
Hi,Which architecture is this? Armhf (32-bit)? Also, which operating system are you using?

**Reply by ZeroGravityByUs** [reply]
Hi!  
  
Sorry, my bad.  
I have a Raspberry Pi 3 and as far as I know it's a 32-bit architecture.   
  
Details below. Many thanks for looking in to it.

**cat /sys/firmware/devicetree/base/model**
```  
Raspberry Pi 3 Model B Rev 1.2
```  
  
**cat /etc/os-release**
```  
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"  
NAME="Raspbian GNU/Linux"  
VERSION_ID="10"  
VERSION="10 (buster)"  
VERSION_CODENAME=buster  
ID=raspbian  
ID_LIKE=debian  
HOME_URL="http://www.raspbian.org/"  
SUPPORT_URL="http://www.raspbian.org/RaspbianForums"  
BUG_REPORT_URL="http://www.raspbian.org/RaspbianBugs"  
```  
  
**cat /proc/version**
```  
Linux version 5.10.17-v7+ (dom@buildbot) (arm-linux-gnueabihf-gcc-8 (Ubuntu/Linaro 8.4.0-3ubuntu1) 8.4.0, GNU ld (GNU Binutils for Ubuntu) 2.34) #1414 SMP Fri Apr 30 13:18:35 BST 2021
```

**Reply by ZeroGravityByUs** [reply]
Hi again!  
  
Now I found out the answer to your question:  
`uname -m` gives `armv7l`  
  
So I'm on a 32-bit architecture.

Any ideas or suggestions? Thanks.

**Reply by jmgiaever** [reply]
Hi, sorry but I've been away regards to the constitution day in Norway :)

I see that you've decided to use the docker version and glad that it worked for you. It seems like this is a common issue on the Debian versions for the RPi's, related to some memory improvements for these devices.

However, there should be a workaround/fix that I can do. I'm aware that you're not using this version yourself, but if you could be kind and give me the output of the following command, that would be great: 
```
snap --version
```
This way I can improve the package and make it work for others. Thank you in advance.

**Reply by ZeroGravityByUs** [reply]
Hi,

Thanks for your reply!   
  
`snap --version`

````  
snap      2.50
snapd     2.50
series    16
raspbian  10
kernel    5.10.17-v7+
````

**Comment by jmgiaever** [comment, marked as answer]
Hi again,

Looks like the hack for Rasbian/Debian on RPi 32-bit systems is:

**Remove the linking to the unknown libarmmem-file**
1. Log into the shell for zwavejs2mqtt: `sudo snap run --shell zwavejs2mqtt`
2. Then edit the file `/var/lib/snapd/hostfs/etc/ld.so.preload` with `vi /var/lib/snapd/hostfs/etc/ld.so.preload` and comment out the line `/usr/lib/arm-linux-gnueabihf/libarmmem-${PLATFORM}.so` with a `#` in the beginning of this file.
    * For those who's not familiar with VI, press `i` just after you launched the editor to enter insert-mode. Then press `#` to insert the character. Exit insert-mode by pressing `ESC` and quit and save by writing `:wq`.
3. exit the editor and exit the shell with `exit`.

**Add the binary-folder to sudoer**
1. Run `sudo visudo` and add `:/snap/bin` to the end of the `secure_path`, BUT before the closing quote `"`. **Note** the `:`-separator!
2. Make sure `visudo` doesn't complain about semantically errors in your file upon saving.
3. Restart your system with `sudo reboot`.

You're now rid of the error about `/etc/ld.so.preload` and the snap-binaries is now also available for sudoer.

**Reply by ZeroGravityByUs** [reply]
Hello!  
  
Thanks a lot! Fantastic! I just tried it, and it works!  Great stuff! Many thanks!

**Reply by jmgiaever** [reply]
Great! If you could change this post as the answer/solution for this post, that would be great. It would lead the users to the correct solution for the version they choose to install. :)

**Reply by jmgiaever** [reply]
Please note that with the change of the name from zwavejs2mqtt to zwave-js-ui, you have to replace this reference everywhere in the above guide.

**Comment by brianthebald** [comment]
I ran into an issue implementing @jmgiaever 's fix, exactly as someone else describes here: (https://forum.snapcraft.io/t/session-101-scope-is-not-a-snap-cgroup/28193) .  `sudo snap run --shell zwavejs2mqtt` throws an error "/user.slice/user-1000.slice/session-101.scope is not a snap cgroup". This is an issue in snapd up to version 2.54.2 at least. (https://forum.snapcraft.io/t/session-101-scope-is-not-a-snap-cgroup/28193) shows that snapd developer maciek-borzecki committed a fix to the snapd codebase, but until that change makes it into a stable release, you can temporarily get around this error by switching to the edge channel for snapd core. Run `sudo snap refresh core --edge` to grab the edge channel build with the bug fix, then go back to the `sudo snap run --shell zwavejs2mqtt` step and proceed from there.

**Reply by jmgiaever** [reply]
Hi, joachimmg on that forum is me too. 😅 

But great that you linked to it here. Didn't cross my mind to share it.

**Comment by alfakenzo** [comment]
HI guys i idid everything like here posted. I use A RPi4 with bullseye and go through your writes, but i get all the time a error on the Serialport.
 Driver: Failed to open the serial port: Error: Operation not permitted, cannot open /dev/ttyAMA0 (ZW0100) 
The serial port runs before with openhab and i also installed zway to check the hardware. Everything is fine and works but not with zwavejs2mqtt. Maybe somene can help me to fix this
greats

**Reply by robertsLando** [reply]
@alfakenzo are you using snap?

**Reply by alfakenzo** [reply]
Hi, yes I use snap to install zwavejs2mqtt.

**Reply by jmgiaever** [reply]
Hi, are you the guy on slack that I've been talking to, that is using the Razberry HAT ?😄

**Reply by alfakenzo** [reply]
Yes that's me. Same nickname. :)

**Reply by jmgiaever** [reply]
Ah, could barely remember my own, so didn't remember yours 😅

But I'm still working on it. I need to make sure that my card is properly attached and configured, and works. So my initial thought was to compile z2m from source that have access to everything before I spend a whole day investigating on something that doesn't work as intended. 

I'll keep you posted. This is something we I'll do in my spare time.

**Reply by jmgiaever** [reply]
Hi,

Sorry my delayed response. I'm working in tourism and we have the peak now in Tromsø!

I just got back home from 13 hrs of work and tested positive on that darn COVID thing mid-day when I got strong symptoms. 

I will have to wait a few days to continue with this at the moment, but will as soon as possible. Would you be able to use the docker version or something else in the meantime?

**Reply by alfakenzo** [reply]
I can try to use docker. But don't thin it will work too. But I can wait and wish you best.

**Reply by jmgiaever** [reply]
Ok, thanks. I'll report back as soon as I'm on track on something. 🤗

**Reply by alfakenzo** [reply]
Hi, hope you are COVID free and could test something with the razberry.
Greets

**Reply by jmgiaever** [reply]
https://forum.snapcraft.io/t/acessing-serial-device-from-a-snap-package/29052?u=joachimmg

**Reply by alfakenzo** [reply]
Hi i wrote this. But i use a RPI4 maybe you can write a how to also for this. And maybe you can write in in the wiki?
Will try your route to the solution but dont understand how you changed this and plugging the zwavejs2mqtt:serial-port to pi:bt-serial.
Greets

**Reply by alfakenzo** [reply]
@jmgiaever my friend can you tell you set the pi:bt-serial?

**Reply by jmgiaever** [reply]
Did you see my reply in Slack?

**Reply by alfakenzo** [reply]
Now I read it. Can you tell me how I can install it in dev-mode?

**Comment by Koektrommel82** [comment]
Earlier this week I upgraded to Domoticz version 2022.2, didn't make a full backup or image prior to the update, and spent days to figure out how to get my system working again with the new Z-Wave JS UI and Mosquitto MQTT broker. Just want to share some of my learnings to help others who, like me, are not software engineers. 

First tip: Make a full backup before updating Domoticz ;-)
Second tip: Ensure your Z-Wave network is healthy and properly set-up (remove dead notes, clean Domoticz from unused Z-Wave devices, et cetera).

System set-up:
- Raspberry Pi 3 Model B+
- Aeotec Z-Stick Gen5+ 
- Raspbian 10
- Domoticz 2022.2
- Snapd 2.57.6
- zwave-js-ui 8.4.1
- Mosquito 

I followed the instructions on: https://www.domoticz.com/wiki/Zwave-JS-UI. I didn't succeed to get the installation working with Docker and Docker Compose. I managed to get the system working using snapd.

I installed the Mosquitto MQTT broker following the installation instructions on: https://www.domoticz.com/wiki/MQTT (there is a link to another website to install the good old legacy way and test your installation). 

I installed Z-Wave JS UI following the instructions provided on this page. Please note that "zwavejs2mqtt" should be changed to "zwave-js-ui". 

I did not manage to resolve `bash: warning:setlocale:LC_ALL:cannot change locale (en_US.UTF-8)`, but it doesn't seem to make a difference.

I used the following commands:
- `sudo snap set zwave-js-ui server.host=0.0.0.0`
- `sudo snap set system experimental.hotplug=true`

In the Z-Wave JS UI, under Settings->Zwave I configured the USB stick as "by-id": /dev/serial/by-id/usb-0658_0200-if00

Initially, the status of devices automatically updated in Domoticz. But clicking on a switch would generate and error or take a long time to respond. The configuration of my Z-wave network had issues (hence tip two above) and I ended up doing a factory reset for the Z-Wave controller, exclude and include all Z-Wave devices again to make it work. During this process, I noticed the controller occasionally got stuck when an exclusion or inclusion operation didn't finish properly. I rebooted the Raspberry Pi every time this occurred to make it work.

**Comment by BlackRockCity** [comment]
zwave2mqtt can run on an ancient Raspberry Pi model B if you install Raspberry Pi OS Lite 
then install Docker

```
curl -fsSL https://get.docker.com -o get-docker.sh
sudo sh get-docker.sh
sudo usermod -aG docker $USER
newgrp docker
```

then install the latest zwave2mqtt image. 
`docker pull zwavejs/zwavejs2mqtt:latest
`

Don't bother with any other installation method, they will all be dead ends (Armv6l architecture wont support snapd and other problems with compiling from source or using precompiled binaries). 

Get an 8GB or larger SD card

RAM requirements: Docker uses 179MB of 512MB RAM, installation of Docker might spike requirements beyond 256MB? I can attest that 512MB works.

Figure out the location of your USB stick with `lsusb` and `dmesg | grep tty` then
`ls /dev/tty*`

Then put it in the device entry below to run the image:

sudo docker run -d \
  --name zwavejs2mqtt \
  --restart unless-stopped \
  -p 8091:8091 \
  -p 3000:3000 \
  -v $(pwd)/store:/usr/src/app/store \
  --device=/dev/ttyACM0 \
  zwavejs/zwavejs2mqtt:latest

In zwave2mqtt web GUI:

Go to settings>zwave and choose the correct serial port for the USB stick 
Go to settings>home assistant and enable the 'ws server' switch

Then follow the instructions in this video... 
https://www.youtube.com/watch?v=JTUcjQzkC0g
